### ISSUE TITLE
embedded hasMany relationship issues

### ISSUE THREAD

**Issue by jshao17** [firebase/emberfire#220]
I've been playing around with an ember app and using firebase to keep track and manage the presence of players in game rooms. 
My first approach was to use 2 models organized basically like this - 
```javascript
Game = DS.Model.extend({
  players: DS.hasMany('player', { embedded: true })
});
Player = DS.Model.extend({
   game: belongsTo('game')
});
```
When a user visits the URL for the game room I would create a user and also register the onDisconnect() like so in the setupController hook for the route.
```javascript
this.player = controller.store.createRecord('player', {
    game: game
});

var playerRef = gameRef.child('players/' + this.player.id);
playerRef.onDisconnect().remove();

game.save();
```
The tricky part here was that besides removing the player onDisconnect, I also had to cleanup the user when they moved between game rooms so I gave the same route a deactivate hook.
```javascript
deactivate: function() {
    this.player.deleteRecord();
    this.player.save();
}
```
This worked great with a single client. I could join a game room and see my presence and whenever I left the room or disconnected it would cleanup after itself. However, when I added more users simulated by multiple browser sessions I noticed that the player data in ember inspector was not kept in sync across clients and I would get into weird corner cases where it would add a user that has already been deleted. I noticed that adding users pushed the new user data to firebase and in turn to all clients, but removing it through deactivate or onDisconnect only removed the data in that particular client and firebase but didn't propagate down to other clients.

Can someone check if this a bug with embedded hasMany relationships for emberfire? As a workaround I have changed the hasMany relationship to async and it works like a charm, but feel like this could also be made to work using embedded relationships.

**Comment by tstirrat** [comment]
There are some bugs with embedded records. Calling `.save()` on embedded items directly has unexpected results. `async: true` is definitely the way to go until we fix some of the weirdness with embedded.

Can you clarify what you mean by "weird corner cases where it would add a user that has already been deleted"?. Do you mean a previously deleted player shows in the player list on a different client?

A minimal repeatable example would be really helpful in this situation (as a jsbin, or github repo), so that we can test emberfire against it as we make embedded record fixes.

**Comment by jshao17** [comment]
You can use [this repo](https://github.com/jshao17/mafia/tree/4bbf84186868f7ce9f7e5ce6a9f1af404d5fdb07) at the linked point in time to test

**Comment by tstirrat** [comment]
ok, thanks!

**Comment by tstirrat** [comment]
What I think is happening is that a `model.destroyRecord()` is trying to delete the record at the top level of the firebase, rather than acting at the embedded location. This is a known bug #135 

The record does not actually exist in the usual location, so the embedded record is not being properly removed. Can you check if this is part of your problem?

I think you'll get better results making the relationship async and doing something like this:

```js
var gameRef = store.adapterFor('game')._getRef('game').child(gameId);
var playerHasManyLink = gameRef.child('players').child(currentPlayer.get('id'))
playerHasManyLink.onDisconnect().remove();

// to remove the player from a game and then delete it
game.get('players').removeObject(player);
game.save().then(function () {
  return player.destroyRecord();
});
```

#249 should have some fixes for this and make it easier to get the game's firebase ref via `game.ref()`. It should also fix the problem with embedded records so that `save/delete` actually works at the correct location in the firebase tree.

**Comment by tstirrat** [comment]
Directly calling `.deleteRecord()` or `.destroyRecord()` on embedded records is fixed in #249
